### PR TITLE
Use slack user ids

### PR DESCRIPTION
**Pull request by jrmils89**
Fixes issue(s) #157 

Changes proposed in this pull request:

- Save slack user id when the employee is created
- Check for missing slack user ids when sending a message, and add them if it can
- Look up the channel id (used to post messages) by slack user id instead of slack username
- Log sent message errors if slack channel for a username can't be found (indicating a bad username) at some point along the way
- DB migration and test updates

@jessieay @nicoleslaw

### CODE REVIEW

File: app/services/message_sender.rb [hunk 1]
```diff
@@ -9,13 +9,20 @@ def initialize(employee, message)
...
+
     if employee.slack_channel_id.nil?
-      channel_id = SlackChannelIdFinder.new(employee.slack_username, client).run
+      channel_id = SlackChannelIdFinder.new(employee.slack_user_id, client).run
       employee.slack_channel_id = channel_id
       employee.save
     end
 
-    if employee.slack_channel_id
+    if !employee.slack_channel_id.nil?
```
Comment (jessieay) [inline app/services/message_sender.rb]: how about simplifying to `if employee.slack_channel_id.present?` ?

File: app/services/slack_user_finder.rb [hunk 2]
```diff
@@ -3,6 +3,12 @@ def existing_user?
     !slack_user.nil?
   end
 
+  def user_id
+    if !slack_user.nil?
```
Comment (jessieay) [inline app/services/slack_user_finder.rb]: what about `if slack_user.present?` ?

File: db/migrate/20160430175239_add_slack_user_id_to_employee.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddSlackUserIdToEmployee < ActiveRecord::Migration
+  def change
+    add_column :employees, :slack_user_id, :string
```
Comment (jessieay) [inline db/migrate/20160430175239_add_slack_user_id_to_employee.rb]: nice! can we add an index on this too? Looks like we're querying on it.

File: app/models/employee.rb [hunk 4]
```diff
@@ -26,4 +26,11 @@ def validate_slack_username_in_org
       )
     end
   end
+
+  def add_slack_user_id_to_employee
+    user_id = EmployeeFinder.new(slack_username).slack_user_id
+    if !user_id.nil?
```
Comment (jessieay) [inline app/models/employee.rb]: how about ` if user_id.present?`

File: app/controllers/employees_controller.rb [hunk 5]
```diff
@@ -33,6 +35,7 @@ def update
     @employee = Employee.find(params[:id])
     @employee.slack_username = params[:employee][:slack_username]
     @employee.validate_slack_username_in_org
+    @employee.add_slack_user_id_to_employee
```
Comment (jessieay) [inline app/controllers/employees_controller.rb]: wondering if we should move some of this logic to its own class. Feels like the controller is getting bigger and bigger. Doesn't need to happen now, but if you were up for it that would be cool.

File: app/services/message_sender.rb [hunk 6]
```diff
@@ -9,13 +9,20 @@ def initialize(employee, message)
   def run
     configure_slack
 
+    user_id = EmployeeFinder.new(employee.slack_username).slack_user_id
+
+    if employee.slack_user_id.nil? && !user_id.nil?
```
Comment (jessieay) [inline app/services/message_sender.rb]: what about: `if employee.slack_user_id.nil? && user_id.present?`

File: app/services/message_sender.rb [hunk 7]
```diff
@@ -9,13 +9,20 @@ def initialize(employee, message)
   def run
     configure_slack
 
+    user_id = EmployeeFinder.new(employee.slack_username).slack_user_id
```
Comment (jessieay) [inline app/services/message_sender.rb]: maybe we should call this var `slack_user_id` for clarity

File: app/services/message_sender.rb [hunk 8]
```diff
@@ -30,7 +37,14 @@ def run
           error: error,
         )
       end
+    else
+      create_sent_scheduled_message(
+        employee: employee,
+        scheduled_message: message,
+        error: StandardError.new("Was unable to find a slack channel for this username"),
```
Comment (jessieay) [inline app/services/message_sender.rb]: oh this is an interesting way to surface this error. very nice! Wondering if we should be more specific with the error message. Something like: `"unable to find slack user with name #{employee.slack_username} and slack user id #{employee.slack_user_id}"` -- I think having that info in the error msg would be very helpful in debugging!

**Comment by jessieay** [comment]
This looks awesome - just a few small comments from me. Once they are addressed, looks good for squash n merge!

File: app/controllers/employees_controller.rb (hunk 5 above)
Comment (jrmils89) [inline app/controllers/employees_controller.rb]: Yeah, I agree that it's getting a little bloated. I could make a new issue for this if you want? I'm not sure I have a ton of time to get to this particularly soon

**Comment by jrmils89** [comment]
OK, this should be ready to go now (after I forgot to push the error text test change the first time). Made the changes based on the feedback here (everything except for abstracting out the controller logic...new issue maybe?).

**Comment by jessieay** [comment]
awesome, thanks for making updates! Merged at de8cec44ec78e128fb7c602813a62651f8e303d3

I will try to tackle some controller refactors in near future. Appreciate this elegant solution!! 🌈

**Pull request closed without merging**